control_panel_api/aws.py

```diff
@@ -1,106 +1,109 @@
+    def __init__(self):
+        self.enabled = settings.ENABLED['write_to_cluster']
+        self.client = boto3.client
 
+    def _do(self, service, method, **kwargs):
+        if self.enabled:
+            client = self.client(service)
+            return getattr(client, method)(**kwargs)
 
-def create_role(role_name, assume_role_policy):
-    """Creates IAM role with the given name"""
+    def create_bucket(self, name, region, acl='private'):
+        self._do('s3', 'create_bucket',
+            Bucket=name,
+            ACL=acl,
+            CreateBucketConfiguration={'LocationConstraint': region})
 
-    aws_api_client("iam").create_role(
-        RoleName=role_name,
-        AssumeRolePolicyDocument=json.dumps(assume_role_policy)
-    )
+    def put_bucket_logging(self, name, target_bucket, target_prefix):
+        self._do('s3', 'put_bucket_logging',
+            Bucket=name,
+            BucketLoggingStatus={
+                'LoggingEnabled': {
+                    'TargetBucket': target_bucket,
+                    'TargetPrefix': target_prefix
+                }
+            })
 
+    def create_policy(self, name, policy_document):
+        self._do('iam', 'create_policy',
+            PolicyName=name,
+            PolicyDocument=json.dumps(policy_document))
 
-def delete_role(role_name):
-    """Delete the given IAM role."""
+    def delete_policy(self, policy_arn):
+        self._do('iam', 'delete_policy', PolicyArn=policy_arn)
 
-    _detach_role_policies(role_name)
-    aws_api_client("iam").delete_role(RoleName=role_name)
+    def create_role(self, role_name, assume_role_policy):
+        """Creates IAM role with the given name"""
 
+        self._do('iam', 'create_role',
+            RoleName=role_name,
+            AssumeRolePolicyDocument=json.dumps(assume_role_policy))
 
-def _detach_role_policies(role_name):
-    """Detaches all the policies from the given role"""
+    def delete_role(self, role_name):
+        """Delete the given IAM role."""
 
-    policies = aws_api_client("iam").list_attached_role_policies(RoleName=role_name)
-    for policy in policies["AttachedPolicies"]:
-        detach_policy_from_role(
-            role_name=role_name,
-            policy_arn=policy["PolicyArn"],
-        )
+        self._detach_role_policies(role_name)
+        self._do('iam', 'delete_role', RoleName=role_name)
 
+    def _detach_role_policies(self, role_name):
+        """Detaches all the policies from the given role"""
 
-def detach_policy_from_entities(policy_arn):
-    """Get all entities to which policy is attached first then call separate detach operations"""
-    entities = aws_api_client("iam").list_entities_for_policy(PolicyArn=policy_arn)
+        policies = self._do(
+            'iam', 'list_attached_role_policies', RoleName=role_name)
 
-    for role in entities["PolicyRoles"]:
-        detach_policy_from_role(policy_arn, role["RoleName"])
+        if policies:
```

Comment:
Is this `if` statement necessary? Clearly if the list is empty the `for` loop would be a no-op.

Reply:
It is necessary, because `policies` is a dict. The `for` loops through `policies["AttachedPolicies"]` - accessing the `"AttachedPolicies"` item will raise a `KeyError` if `policies` is `None`.